src/main/java/com/transferwise/openbanking/client/api/common/BaseClient.java

```diff
@@ -0,0 +1,46 @@
+package com.transferwise.openbanking.client.api.common;
+
+import com.transferwise.openbanking.client.error.ApiCallException;
+import com.transferwise.openbanking.client.json.JsonConverter;
+import com.transferwise.openbanking.client.json.JsonReadException;
+import lombok.AccessLevel;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.client.RestClientException;
+import org.springframework.web.client.RestClientResponseException;
+import org.springframework.web.client.RestOperations;
+
+/**
+ * Base class for all API clients, containing common functionality.
+ */
+@RequiredArgsConstructor(access = AccessLevel.PROTECTED)
+@Slf4j
+public class BaseClient {
+
+    protected final RestOperations restOperations;
+    protected final JsonConverter jsonConverter;
+
+    protected <T, U> ApiResponse<T, U> mapClientExceptionWithResponse(RestClientResponseException e,
+                                                                      Class<U> errorResponseType) {
+        String responseBody = e.getResponseBodyAsString();
+        U parsedFailureResponse = null;
+        try {
+            parsedFailureResponse = jsonConverter.readValue(responseBody, errorResponseType);
+        } catch (JsonReadException jsonReadException) {
+            log.info("Unable to parse failure response body JSON '{}'", jsonReadException.getJson(), jsonReadException);
```

Comment:
Not logging as a warning or error because ASPSPs regularly return error responses which aren't using the structured response format, e.g. returning HTML or message strings instead of JSON, so it would be a bit spammy to do so.